Iterative PopPUNK updates (#259)
bacpop/PopPUNK: docs/poppunk_iterate.rst

named ``<db>/<db>_boundary<n>_clusters.csv``, where ``<n>`` is the boundary iteration number 
(continuous integers increasing from zero). Clusters must contain at least two samples.

The ``poppunk_iterate.py`` script performs the following steps:

1. Starting from the most specific clusters (nearest the origin), it will iteratively add new clusters which are either:
   
	a. Total new clusters
	b. Subsets of existing clusters
	c. Existing clusters are subset of the new cluster.

2. Remove duplicate clusters.
3. Calculate average core distance within this cluster set.
4. Create a tree by nesting smaller clusters within larger clusters whey are subsets of.
5. Output the combined clusters, average core distances, and tree.
6. Cut this tree to pick a set of clusters under a similarity given by a ``--cutoff``.

Step-by-Step Tutorial
---------------------

**Step 1: Sketching (--create-db)**

First, use ``poppunk --create-db`` to sketch input data and calculate distances between samples. 
For the details, refer to :doc:`sketching`. To run this, using the folowing command::
	
	poppunk --create-db --r-files rlist.txt --output <database> --threads 8

> **Step 2: Initial model fittinig (--fit-model)**

Reviewer: ```suggestion
**Step 2: Initial model fitting (--fit-model)**
```

In iterative-PopPUNK, a universal model recommended for fitting the initial model is GMM with 2 or 3 componnets (K=2 or K=3), 
because more datasets can be analysed using this setting (DBSCAN fits sometimes fail to coverage). 
The details for model fitting can be found in :doc:`Fitting new models`. To run this, using the following command::

	poppunk --fit-model bgmm --K 2 --ref-db <database> --output <database> --threads 16

**Step 3: Multi-level clustering by Moving Decision Boundary Iteratively (--fit-model refine --multi-boundary)**

After fitting the initial model with GMM (with 2 or 3 components), refine it by moving the decision boundary 
to mulitple poistions between the origin and the combined decision boundary using the ``--multi-boundary`` option. 

> To expand within-strain component, use ``--neg-shift``. Details can be found in :doc:`Fitting new models` (the **refine** section). 

Reviewer: ```suggestion
To expand within-strain component, use ``--neg-shift``. Details can be found in :doc:`model_fitting` (the **refine** section). 
```

To run this, use the following command::

	poppunk --fit-model refine --multi-boundary 30 --ref-db <database> --output <database> --threads 16

> **Step 4: Chossing clusters under given similarity cutoffs (poppunk_iterate.py)**

Reviewer: ```suggestion
**Step 4: Choosing clusters under given similarity cutoffs (poppunk_iterate.py)**
```

With ``<n>`` sets of clusters created in Step 3, **poppunk_iterate.py** is used to assemble a hierachical tree 
of iterative-PopPUNK clusters, while also calculating the average core distance within each cluster set. To run this step, 
use the following command::

	poppunk_iterate.py --db <database> --cutoff 0.2  --output <prefix> --cpus 16

**Outputs:**

- <prefix>_iterate.tree.nwk: this is the hierachical tree of iterative-PopPUNK clusters.
- <prefix>_iterate.clusters.csv: this file contains all cluster sets from ``<n>`` positions, along with their coresponding average core distances
- <prefix>_iterative.cutoff_clusters.csv: this file contains a single set of clusters that meets a specified similarity cutoff.

.. note::

> 	We recommend using same name for <database> in Step1,2,3 (``--ref-db`` and ``--output``) to save the outputs from each step in a same folder

Reviewer: ```suggestion
	We recommend using same name for <database> as in steps 1-3 (``--ref-db`` and ``--output``) to save the outputs from each step in a same folder
```

> .. contents::
>    :local:

Reviewer: Remove
```suggestion
```

    
.. contents::
   :local:
   
**Step 2: Sublineage clustering for clonal groups**

Next, we extract the clonal groups estimated by Iterative-PoPUNK in Step 1 and run Iterative-PopPUNK again for sublineage clustering. 
For example, we use the two largest clonal groups estimated in Step 1 for downstream analysis.

Sublineage clustering for clonal group CG3::

	poppunk --create-db --r-files cg3_rlist.txt --output VP_cg3 --threads 16 --sketch-size 100000
	poppunk --fit-model bgmm --K 2 --ref-db VP_cg3 --output VP_cg3 --threads 16
	poppunk --fit-model refine --ref-db VP_cg3 --output VP_cg3 --multi-boundary 30 --threads 16 
	poppunk_iterate.py --db VP --cpus 16 --cutoff 0.4 --output 0.4_VP_cg3

Sublineage clustering for clonal group CG189::
	
	poppunk --create-db --r-files cg189_rlist.txt --output VP_cg189 --threads 16 --sketch-size 100000
	poppunk --fit-model bgmm --K 2 --ref-db VP_cg189 --output VP_cg189 --threads 16
	poppunk --fit-model refine --ref-db VP_cg189 --output VP_cg189 --multi-boundary 30 --threads 16 
	poppunk_iterate.py --db VP --cpus 16 --cutoff 0.2 --output 0.2_VP_cg189

**Results**:

.. image:: images/vp_ipp_result.png

> 	:scale: 50%

Reviewer: It would be better to resize the original image
